### ISSUE TITLE
RootInjector should append LazyModule's entryComponents [feature request]

### ISSUE THREAD

**Issue by renatoaraujoc** [angular/angular#17168]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
If you have a singleton service that uses the root ComponentFactoryResolver to dynamically create components and you wish to create an entryComponent declared in a lazy module, this will fail because the root injector does not know about it.

**Expected behavior**
This is actually the expect behavior considering that the root injector will not know about entryComponents known by the child injector.

**What is the motivation / use case for changing the behavior?**
The motivation is to leverage the complexity to create dynamic components even though the modular design of Angular is correct.

* **Angular version:** 2.4.9

**Proposed solution:** Even though this goes against the modular design adopted by Angular itself, I think that the complexity provided to create dynamic components could be softened by simply appending the LazyModule's entryComponents to the RootInjector's entryComponents and to avoid flooding the RootInjector, whenever you navigate out (which destroys the LazyModule), the previously injected entryComponents would be erased from the array.

Simply talking:

1. Page is loaded -> RootInjector is created.
2. Issues a navigation which activates a LazyModule -> ChildInjector is created.
3. RootInjector.entryComponents.append(childInjector.entryComponents)
4. Issues a navigation to somewhere else -> LazyModule is no longer necessary, thus destroyed.
5. RootInjector.entryComponents.destroyLastAppendedEntryComponents()

Hope I could suggest something good.

Thank you!

**Comment by istiti** [comment]
+1 this is  must have feature . I have modal which show history of each component between two date.   I have dialogservice with modal squelette in app.modul but this squelette modal append dynamically all components. 
I can't add to my root entryComponents all my module cause I lost benefit of lazyloding,  i think something like entryComponentsRoot() is very usefull to open root injector and put all these dynamic component inside root injector. Any news here?

**Comment by 0x70b1a5** [comment]
+1, I spent hours hunting this bug down only to discover it's impossible to fix without abandoning the modular paradigm. Not a good look!

**Comment by jmcclanahan** [comment]
@0x70b1a5 - it's very possible while keeping things modular. See #14324. Though it would be nice to have this built into the framework.

**Comment by renatoaraujoc** [comment]
I'd seriously love to have this implemented. Even though it breaks the modular design, it would make life 10 timer easier. I leveraged the solution by passing the childInjector to the service and then ask for the ComponentFactoryResolver for the component I want to render. It's "hackish" but works...

**Comment by istiti** [comment]
@renatoaraujoc +1
I do exactly same and code is very weird and not easy to understand/maintain

as said if child injector (from lazyload) is required by some usecases we can keep `entryComponents[]` but I would ask new options like `entryComponentsRoot[]`

 hope we'll receive news asap, i'm at 10% of my app and it is already big  (more than 50 lazyload/module)

**Comment by Agrumas** [comment]
Agree, the current behavior is unexpected and inconvenient especially working with modals in lazy loading modules. I have chosen to use https://github.com/ng-bootstrap/ng-bootstrap modals to avoid dealing with this.

**Comment by istiti** [comment]
apologies to make noise any ETA here? 

cc @hansl @chuckjaz

**Comment by hansl** [comment]
I was just triaging this issue, I don't have context or information. As it goes, having it on the backlog is the first step to resolve, but it might sit there longer.

**Comment by lucasbasquerotto** [comment]
Perhaps a similar approach to what is done now in Angular6 with `Injectables()` using `providedIn: 'root'` could work.

Related #23819
